controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -70,6 +97,82 @@ ZKStream newStream(final String scope, final String name) {
         return new ZKStream(scope, name, storeHelper);
     }
 
+    @Override
+    CompletableFuture<BigLong> getNextCounter() {
+        CompletableFuture<BigLong> future;
+        synchronized (lock) {
+            BigLong next = counter.incrementAndGet();
+            if (next.compareTo(limit.get()) > 0) {
+                // ignore the counter value and after refreshing call getNextCounter
+                future = refreshRangeIfNeeded().thenCompose(x -> getNextCounter());
+            } else {
+                future = CompletableFuture.completedFuture(next);
+            }
+        }
+        return future;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> refreshRangeIfNeeded() {
+        CompletableFuture<Void> refreshFuture;
+        synchronized (lock) {
+            // Ensure that only one background refresh is happening. For this we will reference the future in refreshFutureRef
+            // If reference future ref is not null, we will return the reference to that future.
+            // It is set to null when refresh completes.
+            refreshFuture = this.refreshFutureRef;
+            if (this.refreshFutureRef == null) {
+                // no ongoing refresh, check if refresh is still needed
+                if (counter.get().compareTo(limit.get()) >= 0) {
+                    log.info("refreshing counter range. Current counter is {}. Current limit is {}", counter.get(), limit.get());
+
+                    // Need to refresh counter and limit. Start a new refresh future. We are under lock so no other
+                    // concurrent thread can start the refresh future.
+                    refreshFutureRef = getRefreshFuture()
+                            .exceptionally(e -> {
+                                // if any exception is thrown here, we would want to reset refresh future so that it can be retried.
+                                synchronized (lock) {
+                                    refreshFutureRef = null;
+                                }
+                                log.warn("Exception thrown while trying to refresh transaction counter range", e);
```

Comment:
Is this warn or error? It looks like the controller can't make progress in the case it fails to refresh the range, and if this is correct, it should be error level.

Reply:
It is warning because controller will try all these steps again if a new createTransaction request comes. 
so only the current transcation creation fails.
its problematic if we repeatedly keep getting the same failure over and over.